### PR TITLE
Add moe.clover.mm3d

### PR DESCRIPTION
**Pull request by zturtleman**
[Maverick Model 3D](https://clover.moe/mm3d), a 3D model editor.

### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [ ] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
`--filesystem=host` is used as the application needs to access adjacent files and absolute paths when loading 3D models and writing 3D models may need to write multiple files.
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: appstream.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+index 5afeaa2..d81055e 100644
+--- a/desktop/moe.clover.mm3d.metainfo.xml
++++ b/desktop/moe.clover.mm3d.metainfo.xml
+@@ -17,9 +17,18 @@
+     </p>
+   </description>
+ 
++  <screenshots>
++    <screenshot type="default">
++      <image type="source">https://raw.githubusercontent.com/clover-moe/screenshots/main/mm3d/1.png</image>
```
Comment (bbhtt) [inline appstream.patch:19]: it should be from a tag or commit

File: appstream.patch
Comment (bbhtt) [inline appstream.patch:1]: did you submit this patch?

File: moe.clover.mm3d.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+finish-args:
+  # OpenGL access
+  - --device=dri
+  # Wayland access
+  - --socket=wayland
+  # X11 + XShm access
+  - --share=ipc
+  - --socket=fallback-x11
+  # File access
+  - --filesystem=host
```
Comment (bbhtt) [inline moe.clover.mm3d.yaml:15]: This seems too broad? Does it use filechooser/filechooser portal? Which paths does it need?

File: moe.clover.mm3d.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  - name: mm3d
+    config-opts:
+      # specify where GLU was just installed to
+      # NOTE: "-L/app/lib" is set in LDFLAGS by flatpak-builder but mm3d configure doesn't use it for testing GL LIBS.
+      - "--with-GL-cflags=-L/app/lib"
+      # specify where Qt is in the SDK
+      - "--with-Qt-bin-dir=/bin"
+      - "--with-Qt-include-dir=/include"
+      - "--with-Qt-lib-dir=/lib"
+      - "--with-Qt-translations-dir=/translations"
```
Comment (bbhtt) [inline moe.clover.mm3d.yaml:27]: These relative to `/usr` right? Also what does it need in /usr/bin for Qt?

File: appstream.patch (hunk 1 above)
Comment (zturtleman) [reply appstream.patch:19]: It references a commit now.

File: appstream.patch
Comment (zturtleman) [reply appstream.patch:1]: Yes, both of the patches are upstream.

Comment (bbhtt) [reply appstream.patch:1]: thanks

**Comment by bbhtt** [comment]
bot, build moe.clover.mm3d

File: moe.clover.mm3d.yaml (hunk 3 above)
Comment (zturtleman) [reply moe.clover.mm3d.yaml]: Ah, yeah they're in /usr. There is no `/translations` symlink so using /usr/translations/ probably fixes finding Qt translations. Thanks.

It uses Qt's build tools; uic, moc, rcc, and lrelease.

**Comment by bbhtt** [comment]
aarch64 is failing:

```
paintwidget.cc: In member function ‘virtual void PaintWidget::initializeGL()’:
paintwidget.cc:49:18: error: ‘GL_SMOOTH’ was not declared in this scope; did you mean ‘GL_SHORT’?
   49 |    glShadeModel( GL_SMOOTH );
      |                  ^~~~~~~~~
      |                  GL_SHORT
paintwidget.cc:49:4: error: ‘glShadeModel’ was not declared in this scope
   49 |    glShadeModel( GL_SMOOTH );
      |    ^~~~~~~~~~~~
paintwidget.cc:52:4: error: ‘glClearDepth’ was not declared in this scope; did you mean ‘glClearDepthf’?
   52 |    glClearDepth( 1.0f );
      |    ^~~~~~~~~~~~
      |    glClearDepthf
paintwidget.cc:58:19: error: ‘GL_LIGHT_MODEL_LOCAL_VIEWER’ was not declared in this scope
   58 |    glLightModeli( GL_LIGHT_MODEL_LOCAL_VIEWER, GL_FALSE );
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
paintwidget.cc:58:4: error: ‘glLightModeli’ was not declared in this scope
   58 |    glLightModeli( GL_LIGHT_MODEL_LOCAL_VIEWER, GL_FALSE );
      |    ^~~~~~~~~~~~~
paintwidget.cc:59:15: error: ‘GL_LIGHT0’ was not declared in this scope; did you mean ‘GL_LIGHTEN’?
   59 |    glLightfv( GL_LIGHT0, GL_AMBIENT, ambient );
      |               ^~~~~~~~~
      |               GL_LIGHTEN
paintwidget.cc:59:26: error: ‘GL_AMBIENT’ was not declared in this scope; did you mean ‘GL_INT’?
   59 |    glLightfv( GL_LIGHT0, GL_AMBIENT, ambient );
      |                          ^~~~~~~~~~
      |                          GL_INT
paintwidget.cc:59:4: error: ‘glLightfv’ was not declared in this scope
   59 |    glLightfv( GL_LIGHT0, GL_AMBIENT, ambient );
      |    ^~~~~~~~~
paintwidget.cc:60:26: error: ‘GL_DIFFUSE’ was not declared in this scope
   60 |    glLightfv( GL_LIGHT0, GL_DIFFUSE, diffuse );
      |                          ^~~~~~~~~~
paintwidget.cc:61:26: error: ‘GL_POSITION’ was not declared in this scope; did you mean ‘GL_LOCATION’?
   61 |    glLightfv( GL_LIGHT0, GL_POSITION, position );
      |                          ^~~~~~~~~~~
      |                          GL_LOCATION
paintwidget.cc:64:15: error: ‘GL_LIGHTING’ was not declared in this scope; did you mean ‘GL_LIGHTEN’?
   64 |    glDisable( GL_LIGHTING );
      |               ^~~~~~~~~~~
      |               GL_LIGHTEN
paintwidget.cc: In member function ‘virtual void PaintWidget::resizeGL(int, int)’:
paintwidget.cc:71:18: error: ‘GL_PROJECTION’ was not declared in this scope; did you mean ‘GL_LOCATION’?
   71 |    glMatrixMode( GL_PROJECTION );
      |                  ^~~~~~~~~~~~~
      |                  GL_LOCATION
paintwidget.cc:71:4: error: ‘glMatrixMode’ was not declared in this scope
   71 |    glMatrixMode( GL_PROJECTION );
      |    ^~~~~~~~~~~~
paintwidget.cc:72:4: error: ‘glLoadIdentity’ was not declared in this scope
   72 |    glLoadIdentity( );
      |    ^~~~~~~~~~~~~~
paintwidget.cc:74:38: error: ‘GL_FILL’ was not declared in this scope
   74 |    glPolygonMode( GL_FRONT_AND_BACK, GL_FILL );
      |                                      ^~~~~~~
paintwidget.cc:74:4: error: ‘glPolygonMode’ was not declared in this scope; did you mean ‘glPolygonOffset’?
   74 |    glPolygonMode( GL_FRONT_AND_BACK, GL_FILL );
      |    ^~~~~~~~~~~~~
      |    glPolygonOffset
paintwidget.cc:76:4: error: ‘glOrtho’ was not declared in this scope
   76 |    glOrtho( 0.0, 1.0, 0.0, 1.0, -1.0, 1.0 );
      |    ^~~~~~~
paintwidget.cc:78:18: error: ‘GL_MODELVIEW’ was not declared in this scope
   78 |    glMatrixMode( GL_MODELVIEW );
      |                  ^~~~~~~~~~~~
make[4]: *** [Makefile:765: paintwidget.o] Error 1
make[4]: *** Waiting for unfinished jobs....
mv -f .deps/contextname.Tpo .deps/contextname.Po
mv -f .deps/alignwin.moc.Tpo .deps/alignwin.moc.Po
viewpanel.cc: In member function ‘void ViewPanel::makeViews()’:
viewpanel.cc:178:8: warning: variable ‘height’ set but not used [-Wunused-but-set-variable]
  178 |    int height = 3;
      |        ^~~~~~
```

**Comment by hfiguiere** [comment]
That's because Qt and openGL is as total mess. The SDK on aarch64 is GLES only (can't have both). So when something goes down the GL API instead of sticking to Qt, it happens. Sometime it's just a matter of including the GL headers explicitely.

**Comment by zturtleman** [comment]
The rendering code requires desktop OpenGL. I disabled aarch64 now.

**Comment by zturtleman** [comment]
bbhtt:
>> - --filesystem=host

> This seems too broad? Does it use filechooser/filechooser portal? Which paths does it need?

The application running in a Flatpak respects XDG_CONFIG_HOME and XDG_DATA_HOME for config and it doesn't need any specific paths. Game content development is probably normally in /home/ directory but may also be an external drive in /media/ or /mnt/. There is no standard location for this.

The real problem is that 3D models are made of several files and may be in different directories.

Loading/saving models may need to access to files outside the directory of the file opened and needs the absolute paths in order to keep relative path references when saving at a different location. (See example below.)

If a texture image is not found, the application lists all files in the directory to do a case insensitive search to find textures in order to be compatible with models created on Windows.

These behaviors and probably others do not seem to fit well with read one file or directory of unknown path and write one file or directory to unknown path. I have not actually tried to limit it though.

## Example
Opening a model ("maps/gamelevel.obj") may reference and load files that are outside it's directory.

```
maps/gamelevel.obj -> relative reference to gamelevel.mtl file
maps/gamelevel.mtl -> relative reference to ../textures/ceiling/1.jpg, etc
textures/ceiling/a.jpg
textures/floor/a.jpg
textures/wall/A.jpg
```

mtl and texture image files that are referenced may also be absolute paths instead of relative paths. If texture image is not found ("textures/wall/a.jpg"), it lists all files in directory ("textures/wall/") to find that A.jpg exists and use it even though the model referenced a.jpg.

The texture image paths are converted to absolute paths internally and exporting the model to a different directory ("maps/multiplayer/gamelevel.obj") has (new) relative paths to the images based on the model filename.

```
maps/multiplayer/gamelevel.obj -> also writes "gamelevel.mtl" in the same directory
maps/multiplayer/gamelevel.mtl -> references "../../textures" instead of "../textures" to access the original texture paths
textures/ceiling/a.jpg -> unchanged
textures/floor/a.jpg -> unchanged
textures/wall/A.jpg -> unchanged
```

File: moe.clover.mm3d.yaml (hunk 2 above)
Comment (zturtleman) [reply moe.clover.mm3d.yaml:15]: I wrote a long response so I posted in the main comments at https://github.com/flathub/flathub/pull/4833#issuecomment-1878058690.

**Comment by barthalion** [comment]
bot, build moe.clover.mm3d

**Pull request closed without merging**